File: Sunrin-ICPC-2021/Final-A/FA.java
import java.util.Scanner;

public class Main{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), inf = 1000000000;
        int[] a = new int[n+1];
        int[][] dp = new int[n+1][4];
        int[][] prv = new int[n+1][4];

        for(int i=1; i<=n; i++) a[i] = sc.nextInt();
        for(int i=1; i<=n; i++) for(int j=0; j<=3; j++) dp[i][j] = inf;

        dp[0][0] = 1;
        for(int i=1; i<=n; i++){
            for(int j=0; j<=3; j++){
                if(dp[i-1][j] <= a[i] && dp[i][j] > a[i]){
                    dp[i][j] = a[i];
                    prv[i][j] = j;
                }
                if(j > 0 && dp[i][j] > dp[i-1][j-1]){
                    dp[i][j] = dp[i-1][j-1];
                    prv[i][j] = j-1;
                }
            }
        }

        int isFinished = 0;
        for(int ti=0; ti<=3; ti++){
            if(isFinished == 0 && dp[n][ti] != inf){
                System.out.printf("YES\n%d\n", ti);
                for(int i=n, j=ti; i>=1; j=prv[i][j], i--){
                    if(prv[i][j] == j-1){
                        System.out.printf("%d %d\n", i, dp[i][j]);
                    }
                }
                isFinished = 1;
            }
        }

        if(isFinished == 0){
            System.out.println("NO");
        }
    }
}

File: Sunrin-ICPC-2021/Final-A/FA.py
n = int(input())
a = list(map(int, input().split()))
a.insert(0, 0)

inf = 1000000000

dp = [[inf for col in range(4)] for row in range(n+1)]
prv = [[0 for col in range(4)] for row in range(n+1)]

dp[0][0] = 1
for i in range(1, n+1):
    for j in range(0, 4):
        if dp[i-1][j] <= a[i] and dp[i][j] > a[i]:
            dp[i][j] = a[i]
            prv[i][j] = j
        if j > 0 and dp[i][j] > dp[i-1][j-1]:
            dp[i][j] = dp[i-1][j-1]
            prv[i][j] = j-1

isFinished = False
for ti in range(0, 4):
    if dp[n][ti] == inf:
        continue
    print("YES")
    print(ti)
    i, j = n, ti
    while i > 0:
        if prv[i][j] == j-1:
            print(i, dp[i][j])
        j = prv[i][j]
        i -= 1
    isFinished = True
    break

if not isFinished:
    print("NO")